### PR TITLE
 feat(snapshots): write `Receipt` to snapshots during `ExecutionStage`

### PR DESCRIPTION
**Pull request by joshieDo**
**attention: it only writes to static files if there's no pruning/configuration enabled for receipts (includes `--full` which only includes deposit contract receipts). Meaning: it keeps them on database**

on top of #5733

### CODE REVIEW

File: crates/storage/provider/src/providers/database/provider.rs [hunk 1]
```diff
@@ -2369,7 +2369,7 @@ impl<TX: DbTxMut + DbTx> BlockWriter for DatabaseProvider<TX> {
             .cursor_read::<tables::TransactionBlock>()?
             .last()?
             .map(|(n, _)| n + 1)
-            .unwrap_or_default();
+            .unwrap_or(1);
```
Comment (joshieDo) [inline crates/storage/provider/src/providers/database/provider.rs:2372]: `tx_num` starts at 1, not 0

File: crates/storage/provider/src/providers/database/mod.rs
Comment (joshieDo) [inline crates/storage/provider/src/providers/database/mod.rs:1]: Changing `insert_block` (previous comment) meant changing these tests to match the live assumption of `tx_num` starting at 1.

**Comment by joshieDo** [comment]
Testing on holesky, but rfr.

File: crates/storage/provider/src/providers/database/mod.rs
Comment (joshieDo) [reply crates/storage/provider/src/providers/database/mod.rs:1]: reverting, since this was my wrong assumption

File: crates/storage/provider/src/providers/database/provider.rs (hunk 1 above)
Comment (joshieDo) [reply crates/storage/provider/src/providers/database/provider.rs:2372]: reverting, since this was my wrong assumption

File: crates/stages/src/stages/execution.rs [hunk 2]
```diff
@@ -418,17 +424,33 @@ impl<EF: ExecutorFactory, DB: Database> Stage<DB> for ExecutionStage<EF> {
...
-
-        while let Some(Ok((tx_number, receipt))) = reverse_walker.next() {
-            if tx_number < first_tx_num {
-                break
+        match maybe_snapshotter(&self.prune_modes, provider, *range.start())? {
+            Some(_) => {
+                // `maybe_snapshotter` already prunes receipts if it detects that there are more
+                // elements in the static files than expected.
+                if let Some(stage_checkpoint) = stage_checkpoint.as_mut() {
+                    for block_number in range {
```
Comment (onbjerg) [inline crates/stages/src/stages/execution.rs:432]: this is to "unwind" the stage progress?

File: crates/stages/src/stages/execution.rs [hunk 3]
```diff
@@ -418,17 +424,33 @@ impl<EF: ExecutorFactory, DB: Database> Stage<DB> for ExecutionStage<EF> {
...
+                        stage_checkpoint.progress.processed -= provider
+                            .block_by_number(block_number)?
+                            .ok_or_else(|| ProviderError::BlockNotFound(block_number.into()))?
+                            .gas_used;
+                    }
+                }
             }
-            reverse_walker.delete_current()?;
+            None => {
```
Comment (onbjerg) [inline crates/stages/src/stages/execution.rs:440]: i'd like if we could either make `maybe_snapshotter` more verbose in its naming, split it out into two separate functions, or add some more comments here. it is a bit hard to follow what exactly goes on since `maybe_snapshotter` (as a name) doesn't say much, but seems to do a lot?

**Review by onbjerg — CHANGES_REQUESTED** [review]

File: crates/stages/src/stages/execution.rs (hunk 2 above)
Comment (joshieDo) [reply crates/stages/src/stages/execution.rs:435]: added more context, lmk if it makes sense

File: crates/stages/src/stages/execution.rs (hunk 3 above)
Comment (joshieDo) [reply crates/stages/src/stages/execution.rs:443]: changed  the name and added context, lmk if it makes sense

File: crates/stages/src/stages/execution.rs [hunk 4]
```diff
@@ -121,6 +123,10 @@ impl<EF: ExecutorFactory> ExecutionStage<EF> {
         let max_block = input.target();
         let prune_modes = self.adjust_prune_modes(provider, start_block, max_block)?;
 
+        // We only append receipts to static files if there is no pruning or filtering configured
+        // for receipts.
+        let snapshotter = try_prepare_snapshotter(&self.prune_modes, provider, start_block)?;
```
Comment (mattsse) [inline crates/stages/src/stages/execution.rs:128]: this function is standalone for testing reasons?

then this makes sense, though I'd also add a fn `try_prepare_snapshotter(&self` if possible

File: crates/stages/src/stages/execution.rs [hunk 5]
```diff
@@ -421,17 +427,39 @@ impl<EF: ExecutorFactory, DB: Database> Stage<DB> for ExecutionStage<EF> {
...
+                while let Some(Ok((tx_number, receipt))) = reverse_walker.next() {
+                    if tx_number < first_tx_num {
+                        break
+                    }
+                    reverse_walker.delete_current()?;
 
-            if let Some(stage_checkpoint) = stage_checkpoint.as_mut() {
-                stage_checkpoint.progress.processed -= receipt.cumulative_gas_used;
+                    if let Some(stage_checkpoint) = stage_checkpoint.as_mut() {
+                        stage_checkpoint.progress.processed -= receipt.cumulative_gas_used;
```
Comment (mattsse) [inline crates/stages/src/stages/execution.rs:460]: why does this `-=`?

File: crates/storage/provider/src/bundle_state/bundle_state_with_receipts.rs [hunk 6]
```diff
@@ -285,15 +285,18 @@ impl BundleStateWithReceipts {
...
 
-    /// Write bundle state to database.
+    /// Write bundle state to database and receipts to either database or static files if
+    /// `snapshotter` is `Some`. It should be none if there is any kind of pruning/filtering over
+    /// the receipts.
     ///
     /// `omit_changed_check` should be set to true of bundle has some of it data
     /// detached, This would make some original values not known.
-    pub fn write_to_db<TX: DbTxMut + DbTx>(
+    pub fn write_to_storage<TX: DbTxMut + DbTx>(
```
Comment (mattsse) [inline crates/storage/provider/src/bundle_state/bundle_state_with_receipts.rs:294]: I'd prefer a where clause instead

File: crates/storage/provider/src/bundle_state/bundle_state_with_receipts.rs [hunk 7]
```diff
@@ -302,16 +305,28 @@ impl BundleStateWithReceipts {
         let mut bodies_cursor = tx.cursor_read::<tables::BlockBodyIndices>()?;
         let mut receipts_cursor = tx.cursor_write::<tables::Receipts>()?;
 
-        for (idx, receipts) in self.receipts.into_iter().enumerate() {
-            if !receipts.is_empty() {
-                let block_number = self.first_block + idx as u64;
-                let (_, body_indices) =
-                    bodies_cursor.seek_exact(block_number)?.unwrap_or_else(|| {
+        let mut body_indices = |block_number: BlockNumber| {
+            Ok::<_,DatabaseError>(bodies_cursor.seek_exact(block_number)?.unwrap_or_else(|| {
```
Comment (mattsse) [inline crates/storage/provider/src/bundle_state/bundle_state_with_receipts.rs:309]: this large Ok looks very strange
can you do this in two steps?

File: crates/storage/provider/src/providers/database/provider.rs [hunk 8]
```diff
@@ -108,9 +108,9 @@ pub struct DatabaseProvider<TX> {
 }
 
 impl<TX> DatabaseProvider<TX> {
-    /// Returns snapshot provider
-    pub fn snapshot_provider(&self) -> Option<Arc<SnapshotProvider>> {
-        self.snapshot_provider.clone()
+    /// Returns a snapshot provider reference
+    pub fn snapshot_provider(&self) -> Option<&Arc<SnapshotProvider>> {
```
Comment (mattsse) [inline crates/storage/provider/src/providers/database/provider.rs:112]: can just clone here instead of &Arc or use deref

File: crates/storage/provider/src/providers/snapshot/manager.rs [hunk 9]
```diff
@@ -264,6 +264,13 @@ impl SnapshotProvider {
                             index.insert(tx_end, current_block_range.clone());
                         })
                         .or_insert_with(|| BTreeMap::from([(tx_end, current_block_range)]));
+                } else if let Some(1) = tx_index.get(&segment).map(|index| index.len()) {
+                    // Only happens if we unwind all the txes/receipts from the first static file.
```
Comment (mattsse) [inline crates/storage/provider/src/providers/snapshot/manager.rs:268]:
```suggestion
                    // Only happens if we unwind all the txs/receipts from the first static file.
```

**Review by mattsse — CHANGES_REQUESTED** [review]
few nits

File: crates/stages/src/stages/execution.rs (hunk 5 above)
Comment (joshieDo) [reply crates/stages/src/stages/execution.rs:460]: because this is part of the unwind and we need to set the total gas progress back .

it already existed beforehand, this is my understanding

File: crates/stages/src/stages/execution.rs (hunk 4 above)
Comment (joshieDo) [reply crates/stages/src/stages/execution.rs]: i guess this was doing too much and not very clear
* check if `snapshotter` is to be used. - put it outside of the function
* consistency check of the static file data (eg. do i have more receipts than database expects? then unwind static files)

added more docs that hopefully clarify

File: crates/storage/provider/src/providers/database/provider.rs (hunk 8 above)
Comment (joshieDo) [reply crates/storage/provider/src/providers/database/provider.rs:112]: We discussed externally that implementing it this way is challenging due to lifetime dependencies in `prepare_snapshotter`. The method `SnapshotProvider::get_writer` returns a mutable reference to a `DashMap `entry type (`SnapshotProviderRW`), which also holds an `Arc<SnapshotProvider>` of the same provider. Attempting to clone or use dereference in this context leads to the error of returning an object with a reference to a local object.

It was further to rename `SnapshotProvider` to `SnapshotProviderInner` and redefine `SnapshotProvider`as an `Arc<SnapshotProviderInner>`, since the usage of `SnapshotProvider` is always accompanied by an `Arc` over it.

**Review by mattsse — APPROVED** [review]
lgtm

File: crates/stages/src/stages/execution.rs [hunk 10]
```diff
@@ -121,6 +123,12 @@ impl<EF: ExecutorFactory> ExecutionStage<EF> {
         let max_block = input.target();
         let prune_modes = self.adjust_prune_modes(provider, start_block, max_block)?;
 
+        // If there is any receipt pruning or filtering, then don't use static files for Receipts.
```
Comment (shekhirin) [inline crates/stages/src/stages/execution.rs:126]: nit
```suggestion
        // We only use static files for Receipts, if there is no receipt pruning of any kind.
```

File: crates/storage/provider/src/providers/snapshot/manager.rs [hunk 11]
```diff
@@ -264,6 +264,13 @@ impl SnapshotProvider {
                             index.insert(tx_end, current_block_range.clone());
                         })
                         .or_insert_with(|| BTreeMap::from([(tx_end, current_block_range)]));
+                } else if let Some(1) = tx_index.get(&segment).map(|index| index.len()) {
+                    // Only happens if we unwind all the txs/receipts from the first static file.
+                    // Should only happen in test scenarios.
```
Comment (shekhirin) [inline crates/storage/provider/src/providers/snapshot/manager.rs:269]: hmm, isn't it possible that we have index of length one during the normal node operation?

**Review by shekhirin — APPROVED** [review]

File: crates/storage/provider/src/providers/snapshot/manager.rs (hunk 11 above)
Comment (joshieDo) [reply crates/storage/provider/src/providers/snapshot/manager.rs:269]: i may be missing something, but i think highly unlikely...
